kv/kvserver/rangefeed: TestUnbufferedRegOnCatchUpSwitchOver failed

kv/kvserver/rangefeed.TestUnbufferedRegOnCatchUpSwitchOver [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20428750?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20428750?buildTab=artifacts#/) on master @ [eb8c0d98da22725422f12eac0f757772de9f0e00](https://github.com/cockroachdb/cockroach/commits/eb8c0d98da22725422f12eac0f757772de9f0e00):


```
=== RUN   TestUnbufferedRegOnCatchUpSwitchOver
--- FAIL: TestUnbufferedRegOnCatchUpSwitchOver (45.15s)
=== RUN   TestUnbufferedRegOnCatchUpSwitchOver/disconnect_after_runOutputLoop
I250904 06:46:35.604204 2892 testutils/soon.go:38  [-] 70  SucceedsSoon: expected 12 events, got 17
I250904 06:46:36.605667 2892 testutils/soon.go:38  [-] 71  SucceedsSoon: expected 12 events, got 17
I250904 06:46:37.606787 2892 testutils/soon.go:38  [-] 72  SucceedsSoon: expected 12 events, got 17
I250904 06:46:38.607424 2892 testutils/soon.go:38  [-] 73  SucceedsSoon: expected 12 events, got 17
I250904 06:46:39.608357 2892 testutils/soon.go:38  [-] 74  SucceedsSoon: expected 12 events, got 17
I250904 06:46:40.609426 2892 testutils/soon.go:38  [-] 75  SucceedsSoon: expected 12 events, got 17
I250904 06:46:41.610293 2892 testutils/soon.go:38  [-] 76  SucceedsSoon: expected 12 events, got 17
I250904 06:46:42.610954 2892 testutils/soon.go:38  [-] 77  SucceedsSoon: expected 12 events, got 17
I250904 06:46:43.611244 2892 testutils/soon.go:38  [-] 78  SucceedsSoon: expected 12 events, got 17
I250904 06:46:44.611837 2892 testutils/soon.go:38  [-] 79  SucceedsSoon: expected 12 events, got 17
I250904 06:46:45.612060 2892 testutils/soon.go:38  [-] 80  SucceedsSoon: expected 12 events, got 17
I250904 06:46:46.613063 2892 testutils/soon.go:38  [-] 81  SucceedsSoon: expected 12 events, got 17
I250904 06:46:47.615035 2892 testutils/soon.go:38  [-] 82  SucceedsSoon: expected 12 events, got 17
I250904 06:46:48.616081 2892 testutils/soon.go:38  [-] 83  SucceedsSoon: expected 12 events, got 17
I250904 06:46:49.616914 2892 testutils/soon.go:38  [-] 84  SucceedsSoon: expected 12 events, got 17
I250904 06:46:50.617724 2892 testutils/soon.go:38  [-] 85  SucceedsSoon: expected 12 events, got 17
I250904 06:46:51.618757 2892 testutils/soon.go:38  [-] 86  SucceedsSoon: expected 12 events, got 17
I250904 06:46:52.619375 2892 testutils/soon.go:38  [-] 87  SucceedsSoon: expected 12 events, got 17
I250904 06:46:53.620302 2892 testutils/soon.go:38  [-] 88  SucceedsSoon: expected 12 events, got 17
I250904 06:46:54.621234 2892 testutils/soon.go:38  [-] 89  SucceedsSoon: expected 12 events, got 17
I250904 06:46:55.621674 2892 testutils/soon.go:38  [-] 90  SucceedsSoon: expected 12 events, got 17
I250904 06:46:56.622746 2892 testutils/soon.go:38  [-] 91  SucceedsSoon: expected 12 events, got 17
I250904 06:46:57.623291 2892 testutils/soon.go:38  [-] 92  SucceedsSoon: expected 12 events, got 17
I250904 06:46:58.626687 2892 testutils/soon.go:38  [-] 93  SucceedsSoon: expected 12 events, got 17
I250904 06:46:59.627311 2892 testutils/soon.go:38  [-] 94  SucceedsSoon: expected 12 events, got 17
I250904 06:47:00.628104 2892 testutils/soon.go:38  [-] 95  SucceedsSoon: expected 12 events, got 17
I250904 06:47:01.631921 2892 testutils/soon.go:38  [-] 96  SucceedsSoon: expected 12 events, got 17
I250904 06:47:02.632724 2892 testutils/soon.go:38  [-] 97  SucceedsSoon: expected 12 events, got 17
I250904 06:47:03.633486 2892 testutils/soon.go:38  [-] 98  SucceedsSoon: expected 12 events, got 17
I250904 06:47:04.634366 2892 testutils/soon.go:38  [-] 99  SucceedsSoon: expected 12 events, got 17
I250904 06:47:05.634724 2892 testutils/soon.go:38  [-] 100  SucceedsSoon: expected 12 events, got 17
I250904 06:47:06.635561 2892 testutils/soon.go:38  [-] 101  SucceedsSoon: expected 12 events, got 17
I250904 06:47:07.636806 2892 testutils/soon.go:38  [-] 102  SucceedsSoon: expected 12 events, got 17
I250904 06:47:08.637793 2892 testutils/soon.go:38  [-] 103  SucceedsSoon: expected 12 events, got 17
I250904 06:47:09.638361 2892 testutils/soon.go:38  [-] 104  SucceedsSoon: expected 12 events, got 17
I250904 06:47:10.638508 2892 testutils/soon.go:38  [-] 105  SucceedsSoon: expected 12 events, got 17
I250904 06:47:11.639251 2892 testutils/soon.go:38  [-] 106  SucceedsSoon: expected 12 events, got 17
I250904 06:47:12.639933 2892 testutils/soon.go:38  [-] 107  SucceedsSoon: expected 12 events, got 17
I250904 06:47:13.640953 2892 testutils/soon.go:38  [-] 108  SucceedsSoon: expected 12 events, got 17
I250904 06:47:14.642142 2892 testutils/soon.go:38  [-] 109  SucceedsSoon: expected 12 events, got 17
I250904 06:47:15.643273 2892 testutils/soon.go:38  [-] 110  SucceedsSoon: expected 12 events, got 17
I250904 06:47:16.643922 2892 testutils/soon.go:38  [-] 111  SucceedsSoon: expected 12 events, got 17
    registry_helper_test.go:256: condition failed to evaluate within 45s: from registry_helper_test.go:261: expected 12 events, got 17
--- FAIL: TestUnbufferedRegOnCatchUpSwitchOver/disconnect_after_runOutputLoop (45.13s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUnbufferedRegOnCatchUpSwitchOver.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-54129

--- a/pkg/kv/kvserver/rangefeed/unbuffered_registration_test.go
+++ b/pkg/kv/kvserver/rangefeed/unbuffered_registration_test.go
@@ -359,19 +359,17 @@ func TestUnbufferedRegOnCatchUpSwitchOver(t *testing.T) {
 			defer wg.Done()
 			r.runOutputLoop(ctx, 0)
 		}()
-		capOfBuf := cap(r.mu.catchUpBuf)
-		r.publish(ctx, ev1, &SharedBudgetAllocation{refCount: 1})
-		r.publish(ctx, ev2, &SharedBudgetAllocation{refCount: 1})
-		r.publish(ctx, ev3, &SharedBudgetAllocation{refCount: 1})
-		r.publish(ctx, ev4, &SharedBudgetAllocation{refCount: 1})
-		r.publish(ctx, ev5, &SharedBudgetAllocation{refCount: 1})
+		postCatchUpEvents := []*kvpb.RangeFeedEvent{ev1, ev2, ev3, ev4, ev5}
+		for _, ev := range postCatchUpEvents {
+			r.publish(ctx, ev, &SharedBudgetAllocation{refCount: 1})
+		}
 		catchUpEvents := expEvents(false)
 		wg.Wait()
 
 		require.False(t, r.getOverflowed())
 		require.Nil(t, r.getBuf())
-		s.waitForEventCount(t, capOfBuf+len(catchUpEvents))
+		s.waitForEventCount(t, len(postCatchUpEvents)+len(catchUpEvents))
 		require.Equal(t,
-			[]*kvpb.RangeFeedEvent{ev1, ev2, ev3, ev4, ev5}, s.mu.events[len(catchUpEvents):])
+			postCatchUpEvents, s.mu.events[len(catchUpEvents):])
 	})
 }

